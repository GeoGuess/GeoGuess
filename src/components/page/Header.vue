<template>
    <div>
        <v-app-bar class="header" height="100">
            <router-link to="/">
                <img class="header__logo" src="@/assets/geoguessLogo.png" />
                <img
                    class="header__logo-min"
                    src="@/../public/img/icons/android-icon-72x72.png"
                />
            </router-link>

            <div class="flex-grow-1" />

            <v-app-bar-nav-icon
                class="header__nav-icon"
                @click="menuMobile = !menuMobile"
            ></v-app-bar-nav-icon>
            <nav class="header__nav" :class="{ visible: menuMobile }">
                <v-btn id="historyBtn" text>
                    <router-link to="/history">
                        {{ $t('Home.historyBtn') }}
                    </router-link>
                </v-btn>
                <div class="header__nav__btns">
                    <v-btn id="aboutBtn" text @click="aboutDialog = true">
                        <v-icon size="30"> mdi-help-circle </v-icon>
                    </v-btn>
                    <v-btn text @click="changeStreamerMode(!streamerMode)">
                        <v-icon size="30">
                            mdi-eye{{ streamerMode ? '-off' : '' }}
                        </v-icon>
                    </v-btn>
                    <v-menu>
                        <template v-slot:activator="{ on }">
                            <v-btn id="languageBtn" text v-on="on">
                                <v-icon size="30"> mdi-translate </v-icon>
                            </v-btn>
                        </template>
                        <v-list id="menuLanguage">
                            <v-list-item
                                v-for="(language, index) in languages"
                                :key="index"
                                @click="switchLanguage(language.value)"
                            >
                                <v-list-item-title>
                                    {{ language.text }}
                                </v-list-item-title>
                            </v-list-item>
                        </v-list>
                    </v-menu>
                </div>
            </nav>
            <v-dialog v-model="aboutDialog">
                <About />
            </v-dialog>
        </v-app-bar>

        <v-alert v-if="demoMode" color="#7289DA" dark class="demo-alert">
            <v-row align="center">
                <v-col class="grow">
                    {{ $t('Demo.message') }}
                </v-col>
                <v-col class="shrink">
                    <v-btn target="_blank" href="https://discord.gg/9GXm6RT">
                        <v-icon left> mdi-discord </v-icon>
                        {{ $t('Demo.btn') }}
                    </v-btn>
                </v-col>
            </v-row>
        </v-alert>
        <HeaderAlert />
    </div>
</template>
<script>
import About from '@/components/page/About';
import { languages, RTL_LANGUAGES } from '../../lang';
import { mapActions, mapState } from 'vuex';
import HeaderAlert from './HeaderAlert.vue';

export default {
    components: {
        About,
        HeaderAlert,
    },
    data() {
        return {
            aboutDialog: false,
            languages,
            menuMobile: false,
        };
    },
    computed: {
        ...mapState({
            streamerMode: (state) => state.homeStore.streamerMode,
        }),
        demoMode() {
            return !!process.env.VUE_APP_DEMO_MODE;
        },
    },
    methods: {
        ...mapActions(['setStreamerMode']),
        changeStreamerMode(streamerMode) {
            this.setStreamerMode(streamerMode);
        },
        switchLanguage(language) {
            this.$i18n.locale = language;
            this.$vuetify.lang.current = language;
            this.$vuetify.rtl = RTL_LANGUAGES.includes(language);
            this.saveLanguage(language);
        },
        saveLanguage(language) {
            localStorage.setItem('language', language);
        },
    },
};
</script>
<style lang="scss" scoped>
.header {
    z-index: 1;
    padding: 0 5%;
    background-color: #f1e9d6 !important;
    .header__nav,
    .header__nav__btns {
        display: flex;
        & > div {
            margin: 0 1.5rem;
        }
    }
    .v-btn {
        a {
            text-decoration: none;
            color: initial;
        }
        font-size: 1.2rem;
    }
    .header__logo {
        height: 6rem;
        width: auto;
    }
    .header__logo-min {
        display: none;
    }
    .header__nav-icon {
        visibility: hidden;
    }
}

@media (max-width: 780px) {
    .header {
        .header__logo {
            display: none;
        }
        .header__logo-min {
            display: block;
        }
        .header__nav {
            &:not(.visible) {
                display: none;
            }
            position: absolute;
            top: 6.2rem;
            right: 0rem;
            background: #f1e9d6;
            padding: 1rem;
            box-shadow: 0px 2px 4px -1px rgb(0 0 0 / 20%);
            border-bottom-left-radius: 0.3125rem;
            border-bottom-right-radius: 0.3125rem;
            max-width: 100%;
            flex-direction: row;
            .header__nav__btns {
                margin: 0;
            }
        }
        .header__nav-icon {
            visibility: visible;
        }
    }
}
</style>
