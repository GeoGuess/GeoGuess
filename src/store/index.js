import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);
// Load all modules.
function loadModules() {
	const localContext = require.context('./modules', false, /([a-z_]+)\.js$/i);

	const modules = localContext
		.keys()
		.map((key) => ({ key, name: key.match(/([a-z_]+)(.store)?\.js$/i)[1] }))
		.reduce(
			(m, { key, name }) => ({
				...m,
				[`${name}Store`]: localContext(key).default,
			}),
			{},
		);

	return { context: localContext, modules };
}

const { context, modules } = loadModules();
const store = new Vuex.Store({
    modules,
});

if (module.hot) {
	// Hot reload whenever any module changes.
	module.hot.accept(context.id, () => {
		const { modules } = loadModules();

		store.hotUpdate({
			modules,
		});
	});
}


export default store;