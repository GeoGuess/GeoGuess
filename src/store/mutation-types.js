//
// HOME
//

export const HOME_SET_NAME_GEOJSON = 'HOME_SET_NAME_GEOJSON';
export const HOME_SET_MAP = 'HOME_SET_MAP';
export const HOME_SET_GEOJSON = 'HOME_SET_GEOJSON';
export const HOME_SET_STATUS_GEOJSON = 'HOME_SET_STATUS_GEOJSON';
export const HOME_SET_GEOJSON_ERROR = 'HOME_SET_GEOJSON_ERROR';

export const HOME_SET_LISTS = 'HOME_SET_LISTS';
export const HOME_SET_LISTS_CUSTOMMAPS = 'HOME_SET_LISTS_CUSTOMMAPS';

export const HOME_SET_HISTORY = 'HOME_SET_HISTORY';

export const HOME_SET_STREAMER_MODE = 'HOME_SET_STREAMER_MODE';
//
// GAME (area)
//
export const GAME_SET_AREAS = 'GAME_SET_AREAS';
export const GAME_SET_LOADING_AREAS = 'GAME_SET_LOADING_AREAS';
export const GAME_SET_URL_AREAS = 'GAME_SET_URL_AREAS';

//
// Settings
//
export const SETTINGS_SET_STEP_DIALOG_ROOM = 'SETTINGS_SET_STEP_DIALOG_ROOM';
export const SETTINGS_SET_MODE_DIALOG_ROOM = 'SETTINGS_SET_MODE_DIALOG_ROOM';
export const SETTINGS_SET_OPEN_DIALOG_ROOM = 'SETTINGS_SET_OPEN_DIALOG_ROOM';
export const SETTINGS_SET_ROOM_ERROR = 'SETTINGS_SET_ROOM_ERROR';
export const SETTINGS_SET_ROOM = 'SETTINGS_SET_ROOM';

export const SETTINGS_SET_GAME_SETTINGS = 'SETTINGS_SET_GAME_SETTINGS';
export const SETTINGS_SET_DIFFICULTY = 'SETTINGS_SET_DIFFICULTY';
export const SETTINGS_SET_BBOX = 'SETTINGS_SET_BBOX';
export const SETTINGS_SET_PLAYER_NAME = 'SETTINGS_SET_PLAYER_NAME';
export const SETTINGS_SET_PLAYERS = 'SETTINGS_SET_PLAYERS';
export const SETTINGS_RESET = 'SETTINGS_RESET';

//
// Alert
//
export const ALERT_SET_ALERT = 'ALERT_SET_ALERT';
