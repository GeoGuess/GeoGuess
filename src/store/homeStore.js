import axios from '@/plugins/axios';
import { validURL } from '@/utils';
import { getLocateString, isGeoJSONValid } from '../utils';
import * as MutationTypes from './mutation-types';
import i18n from '../lang';
import IndexedDBService from '../plugins/IndexedDBService';
import { GeoMap, GeoMapCustom } from '../models/GeoMap';

export default {
    state: () => ({
        map: new GeoMapCustom(),
        loadingGeoJson: false,
        errorMessage: null,
        listMaps: [],
        listAreas: [],
        customsMaps: [],
        history: [],
        streamerMode: localStorage.getItem('streamerMode') === 'true',
    }),
    mutations: {
        [MutationTypes.HOME_SET_GEOJSON](state, geojson) {
            const map = new GeoMapCustom();
            map.name = state.map.name;
            map.geojson = geojson;
            state.map = map;
        },
        [MutationTypes.HOME_SET_NAME_GEOJSON](state, name) {
            state.map.name = name;
        },
        [MutationTypes.HOME_SET_MAP](state, map) {
            state.map = map;
        },
        [MutationTypes.HOME_SET_LISTS](state, lists) {
            state.listMaps = lists.maps;
            state.listAreas = lists.areas;
        },
        [MutationTypes.HOME_SET_LISTS_CUSTOMMAPS](state, customsMaps) {
            state.customsMaps = customsMaps;
        },
        [MutationTypes.HOME_SET_HISTORY](state, history) {
            state.history = history;
        },
        [MutationTypes.HOME_SET_STREAMER_MODE](state, streamerMode) {
            state.streamerMode = streamerMode;
            localStorage.setItem('streamerMode', streamerMode);
        },

        [MutationTypes.HOME_SET_STATUS_GEOJSON](state, status) {
            state.loadingGeoJson = status;
        },
        [MutationTypes.HOME_SET_GEOJSON_ERROR](state, error) {
            state.errorMessage = error;
        },
    },

    getters: {
        geoJsonString(state) {
           if (!state.map || !state.map.geojson) {
                return '';
            }
            return JSON.stringify(state.map.geojson, null, 2);
        },
        geoJson(state) {
            return state.map.geojson;
        },
        isValidGeoJson(state) {
           if (!state.map || !state.map.geojson) {
                return null;
            }
            return isGeoJSONValid(state.map.geojson);
        },
        maps(state) {
            return state.customsMaps
                .map((map) => Object.assign(new GeoMapCustom(), map))
                .concat(
                    state.listMaps.map((map) =>
                        Object.assign(new GeoMap(), map)
                    )
                );
        },
        areasList(state) {
            return state.listAreas.map((map) => ({
                ...map,
                nameLocate: getLocateString(map, 'name', i18n.locale),
                descriptionLocate: getLocateString(
                    map,
                    'description',
                    i18n.locale
                ),
            }));
        },
        nbPlaceVisits(state) {
            return state.history.reduce(
                (a, { rounds }) => a + rounds.length,
                0
            );
        },
    },

    actions: {
        setStreamerMode({ dispatch, commit }, value) {
            commit(MutationTypes.HOME_SET_STREAMER_MODE, value);
            dispatch(
                'alertStore/setAlert',
                value && {
                    title: 'Home.streamerModeActivate',
                    subtitle: 'Home.streamerModeDetails',
                    color: 'streamerMode',
                    icon: 'mdi-twitch',
                },
                { root: true }
            );
        },
        loadPlaceGeoJSON({ commit, state }, place) {
            if (place != null && place != '') {
                if (state.loadingGeoJson) {
                    return;
                }
                commit(MutationTypes.HOME_SET_STATUS_GEOJSON, true);

                commit(MutationTypes.HOME_SET_GEOJSON, null);

                axios
                    .get(
                        `https://nominatim.openstreetmap.org/search/${encodeURIComponent(
                            place.toLowerCase()
                        )}?format=geojson&limit=1&polygon_geojson=1`
                    )
                    .then((res) => {
                        if (
                            res &&
                            res.status === 200 &&
                            res.data.features.length > 0
                        ) {
                            let feature = res.data.features[0];
                            commit(MutationTypes.HOME_SET_GEOJSON, feature);
                            return;
                        }
                        commit(
                            MutationTypes.HOME_SET_GEOJSON_ERROR,
                            'No Found Location'
                        );
                    })
                    .finally(() => {
                        commit(MutationTypes.HOME_SET_STATUS_GEOJSON, false);
                    });
            }
        },
        async loadGeoJsonFromUrl({ commit }, url) {
            if (validURL(url)) {
                // if gist url get raw
                if (RegExp('^(https?://)?gist.github.com/').test(url)) {
                    let urlSplit = url.split('/');
                    if (
                        urlSplit.length > 3 &&
                        urlSplit[urlSplit.length - 1] !== 'raw'
                    ) {
                        urlSplit[urlSplit.length - 3] =
                            'gist.githubusercontent.com';
                        urlSplit.push('raw');
                        url = urlSplit.join('/');
                    }
                }
                const geojson = await axios
                    .get(url)
                    .then((res) => {
                        if (res.status === 200 && res.data) {
                            if (typeof res.data === 'object') {
                                return res.data;
                            } else {
                                return JSON.parse(res.data);
                            }
                        }
                    })
                    .catch((err) => {
                        // eslint-disable-next-line no-console
                        console.log(err);
                    });

                commit(MutationTypes.HOME_SET_GEOJSON, geojson);
            }
        },
        setMapLoaded({ commit }, map) {
            commit(MutationTypes.HOME_SET_MAP, map);
        },
        setGeoJson({ commit }, geojson) {
            commit(MutationTypes.HOME_SET_GEOJSON, geojson);
        },
        async saveGeoJson({ state, dispatch }) {
            await state.map.save();
            dispatch('getListMapsCustoms');
            dispatch(
                'alertStore/setAlert',
                {
                    title: 'Home.mapSavedAlert.title',
                    subtitle: 'Home.mapSavedAlert.subtitle',
                },
                { root: true }
            );
        },
        setGeoJsonString({ commit }, geojson) {
            let obj = null;
            if (geojson !== '') {
                obj = JSON.parse(geojson);
            }
            commit(MutationTypes.HOME_SET_GEOJSON, obj);
        },
        async getListMaps({ commit }) {
            const data = await axios
                .get(
                    process.env.VUE_APP_LIST_MAPS_JSON_URL ||
                        'https://maps.geoguess.games/maps.json',
                    {
                        cache: {
                            maxAge: 1000,
                        },
                    }
                )
                .then((res) => res.data);

            commit(MutationTypes.HOME_SET_LISTS, data);
        },
        async getListMapsCustoms({ commit }) {
            const customsMap = await Promise.resolve(
                IndexedDBService.loadDb().then(async () => {
                    return await IndexedDBService.getAllMaps();
                })
            );
            commit(MutationTypes.HOME_SET_LISTS_CUSTOMMAPS, customsMap);
        },
        loadHistory({ commit }) {
            const history = localStorage.getItem('history')
                ? JSON.parse(localStorage.getItem('history'))
                : [];
            commit(MutationTypes.HOME_SET_HISTORY, history);
        },
    },
};
