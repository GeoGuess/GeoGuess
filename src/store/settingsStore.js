import * as MutationTypes from './mutation-types';
import { GAME_MODE, SCORE_MODE } from '../constants';
import i18n from '../lang';
import router from '../router';
import { getMaxDistanceBbox } from '../utils';
import bbox from '@turf/bbox';
import firebase from 'firebase/app';
import 'firebase/database';

export class GameSettings {
    constructor(
        _allPanorama = false,
        _timeLimitation = 0,
        _mode = GAME_MODE.CLASSIC,
        _timeAttack = false,
        _zoomControl = true,
        _moveControl = true,
        _panControl = true,
        _countdown = 0,
        _scoreMode = SCORE_MODE.NORMAL,
        _areaParams = null
    ) {
        this.allPanorama = _allPanorama;
        this.time = _timeLimitation;
        this.modeSelected = _mode;
        this.timeAttackSelected = _timeAttack;
        this.zoomControl = _zoomControl;
        this.moveControl = _moveControl;
        this.panControl = _panControl;
        this.countdown = _countdown;
        this.scoreMode = _scoreMode;
        this.areaParams = _areaParams;
    }
}

export default {
    namespaced: true,
    state: () => ({
        //Dialog
        isOpenDialogRoom: false,
        loadRoom: false,
        currentComponent: 'settingsMap',
        singlePlayer: true,

        // ROOM
        room: null,
        roomName: '',
        roomErrorMessage: null,
        playerNumber: 0,
        // SETTINGS
        gameSettings: new GameSettings(),
        players: [],
        name: localStorage.getItem('playerName') || '',
        invalidName: false,
    }),
    mutations: {
        [MutationTypes.SETTINGS_SET_ROOM](state, roomName) {
            state.room = firebase.database().ref(roomName);
            state.roomName = roomName;
            // Open Modal
            if (!state.isOpenDialogRoom) {
                state.loadRoom = true;
                state.isOpenDialogRoom = true;
            }

            state.room.once('value', (snapshot) => {
                if (snapshot.child('started').val()) {
                    state.roomErrorMessage = i18n.t(
                        'DialogRoom.alreadyStarted'
                    );
                    state.room.off();
                    return;
                }

                const numberOfPlayers = snapshot
                    .child('playerName')
                    .numChildren();
                const playerNumber = numberOfPlayers + 1;

                state.playerNumber = playerNumber;
                const name = state.name === '' ? i18n.t(
                                'CardRoomPlayerName.anonymousPlayerName'
                            ) + playerNumber : state.name;

                state.room.child('playerName/player'+playerNumber).onDisconnect().remove();

                
                if (numberOfPlayers === 0) {
                    // Put the tentative player's name into the room node
                    // So that other player can't enter as the first player while the player decide the name and room size
                    state.room.child('playerName').update(
                        {
                            player1: name,
                        },
                        (error) => {
                            if (!error) {
                                // Put the timestamp the room is created so the expired rooms can be removed by cloud function
                                state.room.update({
                                    createdAt:
                                        firebase.database.ServerValue.TIMESTAMP,
                                });
                                state.loadRoom = false;
                                state.currentComponent = 'settingsMap';
                            }
                        }
                    );
                } else {
                    // Put other player's tentative name
                    state.room
                        .child('playerName/player' + playerNumber)
                        .set(
                            name,
                            (error) => {
                                if (!error) {
                                    state.loadRoom = false;
                                    state.currentComponent = 'playerName';
                                }
                            }
                        );
                }
            });
        },
        [MutationTypes.SETTINGS_SET_ROOM_ERROR](state, error) {
            state.roomErrorMessage = error;
        },
        [MutationTypes.SETTINGS_SET_GAME_SETTINGS](state, settings) {
            if (settings.modeSelected) {
                settings.areaParams = null;
            }
            if (settings.areaParams) {
                settings.modeSelected = GAME_MODE.CUSTOM_AREA;
            }

            state.gameSettings = {
                ...state.gameSettings,
                ...settings,
            };
        },
        [MutationTypes.SETTINGS_SET_DIFFICULTY](state, difficulty){
            state.difficulty = difficulty;
        },
        [MutationTypes.SETTINGS_SET_BBOX](state, bbox){
            state.bboxObj = bbox;
        },
        [MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM](state, open) {
            state.isOpenDialogRoom = open;
        },
        [MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM](state, singlePlayer) {
            state.singlePlayer = singlePlayer;
            state.currentComponent = singlePlayer ? 'settingsMap' : 'roomName';
        },

        [MutationTypes.SETTINGS_SET_STEP_DIALOG_ROOM](state, step) {
            state.currentComponent = step;
        },
        [MutationTypes.SETTINGS_SET_PLAYER_NAME](state, playerName) {
            state.invalidName = state.players.includes(playerName);
            if (!state.invalidName) {
                state.name = playerName;
                state.room
                    .child('playerName/player' + state.playerNumber)
                    .set(playerName);
            }
        },
        [MutationTypes.SETTINGS_SET_PLAYERS](state, players) {
            state.players = players;
        },
        [MutationTypes.SETTINGS_RESET](state) {
            state.room = null;
            state.roomName = '';
            state.playerNumber = 0;
            state.roomErrorMessage = null;
            state.players = [];
            state.gameSettings = new GameSettings();
        },
    },

    getters: {
        areasJson(state) {
            return state.areas;
        },
    },

    actions: {
        closeDialogRoom({ state, commit, dispatch }, cleanRoom = true) {
            commit(MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM, false);

            // Remove the room
            if (state.room != null) {
                state.room.off();
                if (cleanRoom) {
                    if (state.playerNumber === 1) {
                        // Remove the entire node if the player is the first player
                        state.room.remove();
                    } else {
                        // Remove only the player's name node if the player isn't the first player
                        state.room
                            .child('playerName/player' + this.playerNumber)
                            .remove();
                    }
                }
            }

            dispatch('setMapLoaded', new Map(), { root: true });
            commit(MutationTypes.SETTINGS_RESET);
        },
        openDialogRoom({ commit }, isSinglePlayer = true) {
            commit(MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM, isSinglePlayer);
            commit(MutationTypes.SETTINGS_SET_OPEN_DIALOG_ROOM, true);
        },

        searchRoom({ commit, dispatch, state }, roomName) {
            commit(MutationTypes.SETTINGS_SET_MODE_DIALOG_ROOM, false);
            if (roomName == '') {
                commit(
                    MutationTypes.SETTINGS_SET_ROOM_ERROR,
                    i18n.t('DialogRoom.invalidRoomName')
                );
            } else {
                commit(MutationTypes.SETTINGS_SET_ROOM, roomName);
            }

            state.room.on('value', (snapshot) => {
                if (snapshot.child('playerName').exists())
                    state.players = Object.values(
                        snapshot.child('playerName').val()
                    );

                if (
                    state.currentComponent === 'playerName' &&
                    state.playerNumber !== 1 &&
                    snapshot.hasChild('size') &&
                    snapshot.hasChild('streetView')
                ) {
                    dispatch('startGame');
                }
            });
        },
        setSettings({ commit, state, rootState, dispatch }) {
            let difficulty = 2000;
            let bboxObj;
            if (rootState.homeStore.map.geojson) {
                bboxObj = bbox(rootState.homeStore.map.geojson);
                commit(MutationTypes.SETTINGS_SET_BBOX, bboxObj);

                difficulty = getMaxDistanceBbox(bboxObj) / 10;
            }
            commit(MutationTypes.SETTINGS_SET_DIFFICULTY, difficulty);
            if (!state.room) {
                router.push({
                    name: 'street-view',
                    params: {
                        ...state.gameSettings,
                        difficulty,
                        placeGeoJson: rootState.homeStore.map.geojson,
                        bboxObj: bboxObj,
                    },
                });
                dispatch('closeDialogRoom');
            } else {
                state.room.update(
                    {
                        ...state.gameSettings,
                        timeLimitation: state.gameSettings.time,
                        difficulty,
                        ...(rootState.homeStore.map.geojson && {placeGeoJson: rootState.homeStore.map.geojson}),
                        ...(bboxObj && { bboxObj: bboxObj }),
                    },
                    (error) => {
                        if (!error) {
                            commit(
                                MutationTypes.SETTINGS_SET_STEP_DIALOG_ROOM,
                                'playerName'
                            );
                        }
                    }
                );
            }
        },

        setPlayerName({ commit, state }, playerName) {
            if (playerName === '') {
                commit(
                    MutationTypes.SETTINGS_SET_PLAYER_NAME,
                    i18n.t('CardRoomPlayerName.anonymousPlayerName') +' '+ state.playerNumber
                );
            }else{
                localStorage.setItem('playerName', playerName);
                commit(MutationTypes.SETTINGS_SET_PLAYER_NAME, playerName);
            }

        },
        startGame({ state, dispatch, rootState }) {
            let gameParams = {};
            if (state.playerNumber === 1) {
                gameParams = {
                    ...state.gameSettings,
                    difficulty: state.difficulty,
                    placeGeoJson: rootState.homeStore.map.geojson,
                    bboxObj: state.bboxObj,
                };
                // Set flag started
                state.room.update({
                    size: state.players.length,
                    started: true,
                });
                dispatch('startGameMultiplayer', gameParams);
            } else {
                state.room.once('value', (snapshot) => {
                    gameParams = {
                        difficulty: snapshot.child('difficulty').val(),
                        bboxObj: snapshot.child('bboxObj').val(),
                        modeSelected: snapshot.child('modeSelected').val(),
                        timeAttackSelected: snapshot
                            .child('timeAttackSelected')
                            .val(),
                        zoomControl: snapshot.child('zoomControl').val(),
                        moveControl: snapshot.child('moveControl').val(),
                        panControl: snapshot.child('panControl').val(),
                        countdown: snapshot.child('countdown').val(),
                        allPanorama: snapshot.child('allPanorama').val(),
                        scoreMode: snapshot.child('scoreMode').val(),
                        areaParams: snapshot.child('areaParams').val(),
                    };

                    dispatch('startGameMultiplayer', gameParams);
                });
            }
        },
        startGameMultiplayer({ state, rootState, dispatch }, gameParams) {
            // Start the game
            router.push({
                name: 'with-friends',
                params: {
                    ...gameParams,
                    roomName: state.roomName,
                    playerName: state.name,
                    playerNumber: state.playerNumber,
                    placeGeoJson: rootState.homeStore.map.geojson,
                    multiplayer: true,
                },
            });

            dispatch('closeDialogRoom', false);
        },
    },
};
