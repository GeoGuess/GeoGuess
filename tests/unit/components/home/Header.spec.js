import Header from '@/components/page/Header.vue';
import { createLocalVue, mount, shallowMount } from '@vue/test-utils';
import homeStore from '@/store/homeStore';
import * as MutationTypes from '@/store/mutation-types';
import appInit from '../../testutils/appInit';
import Vuex from 'vuex';

const args = appInit(createLocalVue());

describe('Header.vue', () => {
    const setStreamerModeStore = jest.fn();
    let store;
    beforeEach(() => {
        store = new Vuex.Store({
            modules: {
                homeStore: {
                    state: homeStore.state,
                    getters: homeStore.getters,
                    actions: homeStore.actions,
                    mutations: {
                        [MutationTypes.HOME_SET_STREAMER_MODE]: setStreamerModeStore,
                    },
                },
            },
        });
    });
    it('open dialog about', () => {
        const wrapper = mount(Header, {
            ...args,
            store,
        });

        const aboutBtn = wrapper.find('#aboutBtn');
        expect(wrapper.vm.aboutDialog).toEqual(false);
        aboutBtn.trigger('click');
        expect(wrapper.vm.aboutDialog).toEqual(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('switchLanguage method', () => {
        const wrapper = shallowMount(Header, {
            ...args,
            store,
        });
        expect(localStorage.getItem('language')).toEqual('en');
        wrapper.vm.switchLanguage('fr');
        expect(localStorage.getItem('language')).toEqual('fr');
    });

});
