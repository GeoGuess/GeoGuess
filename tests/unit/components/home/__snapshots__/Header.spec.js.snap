// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue changeStreamerMode method 1`] = `
<div>
  <v-app-bar-stub height="100" tag="header" extensionheight="48" src="" value="true" class="header">
    <router-link to="/"><img src="@/assets/geoguessLogo.png" class="header__logo"> <img src="@/../public/img/icons/android-icon-72x72.png" class="header__logo-min"></router-link>
    <div class="flex-grow-1"></div>
    <v-app-bar-nav-icon-stub class="header__nav-icon"></v-app-bar-nav-icon-stub>
    <nav class="header__nav">
      <v-btn-stub tag="button" activeclass="" text="true" type="button" id="historyBtn">
        <router-link to="/history">
          Historique
        </router-link>
      </v-btn-stub>
      <div class="header__nav__btns">
        <v-btn-stub tag="button" activeclass="" text="true" type="button" id="aboutBtn">
          <v-icon-stub size="30"> mdi-help-circle </v-icon-stub>
        </v-btn-stub>
        <v-btn-stub tag="button" activeclass="" text="true" type="button">
          <v-icon-stub size="30">
            mdi-eye
          </v-icon-stub>
        </v-btn-stub>
        <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" origin="top left" transition="v-menu-transition">
          <v-list-stub tag="div" id="menuLanguage">
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                English
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Français
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                čeština
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                русский
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Português (Brasil)
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Deutsch
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Italiano
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Svenska
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                Türkçe
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                עִבְרִית
              </v-list-item-title-stub>
            </v-list-item-stub>
            <v-list-item-stub activeclass="" tag="div">
              <v-list-item-title-stub>
                日本語
              </v-list-item-title-stub>
            </v-list-item-stub>
          </v-list-stub>
        </v-menu-stub>
      </div>
    </nav>
    <v-dialog-stub opendelay="0" closedelay="0" contentclass="" maxwidth="none" origin="center center" retainfocus="true" transition="dialog-transition" width="auto">
      <about-stub></about-stub>
    </v-dialog-stub>
  </v-app-bar-stub>
  <!---->
  <v-alert-stub color="streamerMode" tag="div" transition="slide-x-reverse-transition" closelabel="$vuetify.close" dismissible="true" closeicon="$cancel" icon="mdi-twitch" type="success" id="alertStreamerMode">
    <h4>Mode Streamer activé</h4>
    <p>Les URLs et les noms des salles sont masqués</p>
  </v-alert-stub>
</div>
`;

exports[`Header.vue open dialog about 1`] = `
<div>
  <header class="header v-sheet theme--light v-toolbar v-app-bar" style="height: 100px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;">
    <div class="v-toolbar__content" style="height: 100px;">
      <router-link to="/"><img src="@/assets/geoguessLogo.png" class="header__logo"> <img src="@/../public/img/icons/android-icon-72x72.png" class="header__logo-min"></router-link>
      <div class="flex-grow-1"></div> <button type="button" class="v-app-bar__nav-icon header__nav-icon v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu theme--light"></i></span></button>
      <nav class="header__nav"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default" id="historyBtn"><span class="v-btn__content"><router-link to="/history">
                    HISTORY
                </router-link></span></button>
        <div class="header__nav__btns"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default" id="aboutBtn"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-help-circle theme--light" style="font-size: 30px;"></i></span></button> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-eye theme--light" style="font-size: 30px;"></i></span></button>
          <div class="v-menu"><button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default" id="languageBtn"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-translate theme--light" style="font-size: 30px;"></i></span></button>
            <!---->
          </div>
        </div>
      </nav>
      <div role="dialog" class="v-dialog__container">
        <!---->
      </div>
    </div>
  </header>
  <!---->
  <transition-stub name="slide-x-reverse-transition" id="alertStreamerMode">
    <div role="alert" class="v-alert v-sheet theme--dark streamerMode" style="display: none;">
      <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-twitch theme--dark"></i>
        <div class="v-alert__content">
          <h4>Streamer mode activate</h4>
          <p>URLs and room names are hidden</p>
        </div><button type="button" class="v-alert__dismissible v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--small" aria-label="Close"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close-circle theme--dark"></i></span></button>
      </div>
    </div>
  </transition-stub>
</div>
`;
